net/multi_listen.go

```diff
@@ -0,0 +1,153 @@
+	closed         atomic.Bool
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(ctx context.Context, addresses []string) (net.Listener, error) {
+	return multiListen(
+		ctx,
+		addresses,
+		net.ResolveTCPAddr,
+		func(ctx context.Context, network, address string) (net.Listener, error) {
+			var lc net.ListenConfig
+			return lc.Listen(ctx, network, address)
+		})
+}
+
+func multiListen(
+	ctx context.Context,
+	addresses []string,
+	resolveTCPAddrFunc func(network, address string) (*net.TCPAddr, error),
+	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	ml := &multiListener{
+		connErrIndexCh: make(chan connErrIndexTriplet),
+	}
+
+	for _, address := range addresses {
+		addr, err := resolveTCPAddrFunc("tcp", address)
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		var network string
+		host, _, err := net.SplitHostPort(addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+		switch IPFamilyOf(ParseIPSloppy(host)) {
+		case IPv4:
+			network = "tcp4"
+		case IPv6:
+			network = "tcp6"
+		default:
+			// close all listeners
+			_ = ml.Close()
+			return nil, fmt.Errorf("failed to identify ip family of address '%s", addr.String())
+		}
+
+		l, err := listenFunc(ctx, network, addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		ml.addrs = append(ml.addrs, addr)
+		ml.listeners = append(ml.listeners, l)
+	}
+
+	for i := range ml.listeners {
+		index := i
+		ml.wg.Add(1)
+		// spawn a go routine for every listener to wait for incoming connection requests
+		go func() {
+			defer ml.wg.Done()
+			for {
+				conn, err := ml.listeners[index].Accept()
+				if err != nil && ml.closed.Load() {
+					return
+				}
+				ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
+			}
+		}()
+	}
+	return ml, nil
+}
+
+// Accept is part of net.Listener interface.
+func (ml *multiListener) Accept() (net.Conn, error) {
+	connErrIndex, ok := <-ml.connErrIndexCh
+	if !ok {
+		return nil, fmt.Errorf("use of closed network connection")
+	}
+	// update latestAcceptedIndex with index of the listener which accepted the connection
+	ml.latestAcceptedIndex.Store(int32(connErrIndex.index))
+	return connErrIndex.conn, connErrIndex.err
+
+}
+
+// Close is part of net.Listener interface.
+func (ml *multiListener) Close() error {
+	ml.closed.Store(true)
+	close(ml.connErrIndexCh)
+	for i := range ml.listeners {
+		_ = ml.listeners[i].Close()
+	}
+	ml.wg.Wait()
+	return nil
+}
+
+// Addr is part of net.Listener interface.
+func (ml *multiListener) Addr() net.Addr {
+	return ml.addrs[ml.latestAcceptedIndex.Load()]
```

Comment:
IMO the "latestAcceptedIndex" should be removed. It is unreliable (race) and may lure a caller to actually think it can be used for something sensible. Make a comment instead that the Addr() call _is_ unreliable, and the caller should use conn.LocalAddr() to get the _actual_ local address.
Just return ml.addrs[0].

Reply:
how a mutex instead of atmoic.Bool ?
```go
			for {
			       conn, err := ml.listeners[index].Accept()
			       ml.mux.Lock()
			       if err != nil && ml.closed {
					return
			       }
			       ml.connErrIndexCh <- connErrIndexTriplet{conn: conn, err: err, index: index}
			       ml.mux.Unlock()
			}
```